### ISSUE TITLE
CMake errors : "Could NOT find Sqlite" (Building QtWebkit on Windows)

### ISSUE THREAD

**Issue by kaamui** [annulen/webkit#814]
I'm trying to build qtwebkit on Windows following [this wiki](https://github.com/annulen/webkit/wiki/Building-QtWebKit-on-Windows). 

I'm also trying to adapt the instructions to Visual Studio Community 2019, but with the latest version of msvc officially supported by qt5.12.3, msvc 2017.

I've installed : 

- Visual Studio 2019 Community (+msvc 2017 (v141))
- Qt 5.12.3
- cmake 3.14.4
- Ruby 2.4 + devkit
- ActivePerl 5.26.3
- Python 2.7

```
perl Tools/Scripts/build-webkit --qt --release --cmakeargs="-Wno-dev -G \"Visual Studio 16 2019\" -A \"x64\" -T \"v141\" -DCMAKE_PREFIX_PATH=C:\Qt\5.12.3\msvc2017_64"
```

```
C:\Users\dev\openboard\OpenBoard-ThirdParty\qtwebkit>perl Tools/Scripts/build-webkit --qt --release --cmakeargs="-Wno-dev -G \"Visual Studio 16 2019\" -A \"x64\" -T \"v141\" -DCMAKE_PREFIX_PATH=C:\Qt\5.12.3\msvc2017_64"
Checking Last-Modified date of qtwebkit-libs-win.zip...
Located a file of type application/octet-stream, of size 127044099.
Downloading qtwebkit-libs-win.zip...

https://github.com/Vitallium/qtwebkit-libs-win/releases/download/2015/qtwebkit-libs-win.zip

Installing qtwebkit-libs-win...
Couldn't move qtwebkit-libs-win.headers to C:\Users\dev\openboard\OpenBoard-ThirdParty\qtwebkit\WebKitLibraries\win.
The qtwebkit-libs-win has been sucessfully installed in
 C:\Users\dev\openboard\OpenBoard-ThirdParty\qtwebkit\WebKitLibraries\win
Not searching for unused variables given on the command line.
-- Could NOT find Ruby (missing: RUBY_LIBRARY) (found suitable version "2.4.6", minimum required is "1.9")
-- Looking for include file features.h
-- Looking for include file features.h - not found
CMake Error at C:/Program Files/CMake/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find Sqlite (missing: SQLITE_INCLUDE_DIR SQLITE_LIBRARIES)
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  Source/cmake/FindSqlite.cmake:62 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  Source/cmake/OptionsQt.cmake:437 (find_package)
  Source/cmake/WebKitCommon.cmake:50 (include)
  CMakeLists.txt:137 (include)


-- Configuring incomplete, errors occurred!
See also "C:/Users/dev/openboard/OpenBoard-ThirdParty/qtwebkit/WebKitBuild/Release/CMakeFiles/CMakeOutput.log".
See also "C:/Users/dev/openboard/OpenBoard-ThirdParty/qtwebkit/WebKitBuild/Release/CMakeFiles/CMakeError.log".
```

I've got this error below, but this dependency is never listed on the wiki, so I wonder if it is needed, and if so or not, what I am supposed to do. 

```
Could NOT find Sqlite (missing: SQLITE_INCLUDE_DIR SQLITE_LIBRARIES)
Call Stack (most recent call first):
```

Also, I don't know if  I have to define RUBY_LIBRARY myself or not... (I installed Ruby 2.4 + devkit, and the library is here (but with a strange  'a' extension ("libx64-msvcrt-ruby240.dll.a"))

```
-- Could NOT find Ruby (missing: RUBY_LIBRARY) (found suitable version "2.4.6", minimum required is "1.9")
```

Any ideas that could help me ? :)

**Comment by annulen** [comment]
Please check if dependencies are correctly unpacked to `C:\Users\dev\openboard\OpenBoard-ThirdParty\qtwebkit\WebKitLibraries\win`. This directory should have `include/sqlite3` and `bin64/sqlite3.lib`

>Also, I don't know if I have to define RUBY_LIBRARY myself or not

Ignore this error, WebKit doesn't need RUBY_LIBRARY at all

**Comment by kaamui** [comment]
Ok I didn't pay attention to this message but it seems that the downloaded zip fails to install. 

```
Checking Last-Modified date of qtwebkit-libs-win.zip...
Located a file of type application/octet-stream, of size 127044099.
Downloading qtwebkit-libs-win.zip...

https://github.com/Vitallium/qtwebkit-libs-win/releases/download/2015/qtwebkit-libs-win.zip

Installing qtwebkit-libs-win...
Couldn't move qtwebkit-libs-win.headers to C:\Users\dev\openboard\OpenBoard-ThirdParty\qtwebkit\WebKitLibraries\win.
The qtwebkit-libs-win has been sucessfully installed in
 C:\Users\dev\openboard\OpenBoard-ThirdParty\qtwebkit\WebKitLibraries\win
Not searching for unused variables given on the command line.
```

It actually isn't in my folder. I'm going to add it manually.

```
$ ls -l
total 50
-rw-r--r-- 1 dev 197121  6358 mai   20 10:56 CMakeLists.txt
drwxr-xr-x 1 dev 197121     0 mai   20 10:56 dist/
-rw-r--r-- 1 dev 197121 27346 mai   20 10:56 LICENSE.LGPLv21
-rw-r--r-- 1 dev 197121   722 mai   20 10:56 README.md
drwxr-xr-x 1 dev 197121     0 mai   20 10:56 Source/
drwxr-xr-x 1 dev 197121     0 mai   20 10:56 Tools/
-rw-r--r-- 1 dev 197121   666 mai   20 10:56 WebKit.pro

```

**Comment by kaamui** [comment]
Downloading manually the zip and move it manually to the newly created/expected folders solved it. 

I'm now ready for the next error ! xD

```
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/dev/openboard/OpenBoard-ThirdParty/qtwebkit/WebKitBuild/Release
Microsoft (R) Build Engine version 16.1.76+g14b0a930a7 pour .NET Framework
Copyright (C) Microsoft Corporation. Tous droits réservés.

MSBUILD : error MSB1001: Commutateur inconnu.
Commutateur : -j2

Pour la syntaxe du commutateur, tapez "MSBuild -help"
```

I don't know why -j2 would not be recognized, but I opened the sln file in visual studio and it is compiling so ... I don't know if I should open a new ticket for this issue ?

**Comment by kaamui** [comment]
My bad ... I started trying to build qtwebkit using this repo : 

https://github.com/qt/qtwebkit

and the error "couldn't move ..." above was logical as such directory doesn't exist in this repository. 

For the -j2 switch, I commented this line in the Tools/Scripts/build-webkit script : 

```
# Visual Studio generator doesn't support -j switch
    if (canUseNinja() || !isAnyWindows()) {
        # By default we build using all of the available CPUs.
        # commented below because I have ninja installed but the -j switch is still an issue
        # $makeArgs .= ($makeArgs ? " " : "") . "-j" . numberOfCPUs() if $makeArgs !~ /-j\s*\d+/;
        my $maxCPULoad = maxCPULoad() if $makeArgs !~ /-l\s*\d+\.?\d*/;
        $makeArgs .= " -l" . maxCPULoad() if defined $maxCPULoad;
    }
```

It most probably is a configuration issue with my ninja install... 

Thank you for your time and help, you do a great job !

**Comment by annulen** [comment]
It should be possible to build that repo in this way, however only "conan" instructions were tested in CI. As you may see, our windows build is quite messy. Btw, if you are not planning to work on QtWebKit itself, you should better use ninja generator instead of visual studio one

**Comment by kaamui** [comment]
I also tried the "conan" approach but had issues (maybe too recent build of conan) with it too. 

For example, with conan 1.15.1 x64 this command ... 

```
conan install -f ..\..\qt5\coin\provisioning\qtci-windows-10-x86\conanfiles\qtwebkit.txt -s arch=x86 -s compiler="Visual Studio" -s compiler.version=15 -s compiler.runtime=MD -g cmake
```

... doesn't work anymore as the -f option no longer exists. I think the new usage is to juste remove the -f key but not its value, but I'm not sure. I also had other issues I don't remember. 

I think I'm close to a working solution with the wiki "Build on Windows" (but without ninja and multi-threaded compiling). Still an issue with the qtwebkit libs themselves though as I only got this for now : 

```
$ ls -l WebKitBuild/Release/lib64/
total 492584
-rw-r--r-- 1 dev 197121  15342202 mai   31 10:44 ANGLESupport.lib
-rw-r--r-- 1 dev 197121    270422 mai   31 11:17 brotli.lib
-rw-r--r-- 1 dev 197121   2036766 mai   31 12:40 gtest.lib
-rw-r--r-- 1 dev 197121 477321232 mai   31 11:17 JavaScriptCore.lib
-rw-r--r-- 1 dev 197121    386054 mai   31 12:42 jsc.exp
-rw-r--r-- 1 dev 197121    625054 mai   31 12:42 jsc.lib
-rw-r--r-- 1 dev 197121     91118 mai   31 10:45 LLIntOffsetsExtractor.exp
-rw-r--r-- 1 dev 197121    156206 mai   31 10:45 LLIntOffsetsExtractor.lib
-rw-r--r-- 1 dev 197121     96884 mai   31 12:41 TestWTF.exp
-rw-r--r-- 1 dev 197121    159534 mai   31 12:41 TestWTF.lib
-rw-r--r-- 1 dev 197121    502338 mai   31 11:17 woff2.lib
-rw-r--r-- 1 dev 197121   7390362 mai   31 10:45 WTF.lib
```
The error is related to MSVC 15.8+ compilers : #769 

It seems you addressed this error in the qtwebkit-stable branch but not in the qtwebkit-5.212. Could you merge it in this branch or do I have to work only on qtwebkit-stable (but I'm interested by the recent PR merged (solving the annonying libpng warning) ?

**Comment by annulen** [comment]
Done

**Comment by annulen** [comment]
As for Conan, version 0.24 should work

**Comment by kaamui** [comment]
Thank you,

for the two issues referenced in #769, C2338 is fixed with your merge. 

C2139 error has been fixed Visual Studio 2019 according to [this thread](https://developercommunity.visualstudio.com/content/problem/343550/webkit-an-undefined-class-is-not-allowed-as-an-arg.html), but probably only with MSVC 2019, because with my -T \"v141\" option it is still occurring (only on WebKit2). 

I was hoping to be able to build qtwebkit libs with MSVC 2017 as I'm always scared of external symbols issues, or conflict between libs built with different compilers.. do you think it is safe to build qtwebkit libs with msvc 2019 (64) even if I use msvc2017_64 in qt ?

I might be able to finish generating qtwebkit libs by disabling webkit2 if not, to be honest I'm not sure what I really need to build qtwebkit the way my app needs to. Is there a page or a thread detailing the different options ?

**Comment by annulen** [comment]
>C2139 error has been fixed in Visual Studio 2019 according to this thread, but probably only with MSVC 2019, because with my -T "v141" option it is still occurring (only on WebKit2).

Huh. I've heard about this issue from Sony developers working on WebKit, but I wasn't aware it affects my branch. So far, the best thing they could come with is moving to newer MSVC and not supporting buggy one. Fortunately, MS is now testing their new compiler releases on WebKit code base so it's less likely that we'll have such bugs in future.

However, in your case there simple solution. As you are not using QML API, just build with -DENABLE_WEBKIT2=OFF (or toggle this build option in cmake gui or ccmake)

Note that there are other options which you may want to disable if you don't need related features in your custom build:
```
ENABLE_SPELLCHECK
USE_LIBHYPHEN
USE_WOFF2
ENABLE_VIDEO
ENABLE_WEB_AUDIO
USE_GSTREAMER
ENABLE_GEOLOCATION
ENABLE_DEVICE_ORIENTATION
ENABLE_OPENGL
ENABLE_PRINT_SUPPORT
ENABLE_XSLT
USE_WEBP
ENABLE_TOOLS
ENABLE_TEST_SUPPORT
ENABLE_JIT
ENABLE_INSPECTOR_UI
ENABLE_QT_GESTURE_EVENTS
ENABLE_TOUCH_EVENTS
ENABLE_ICONDATABASE
ENABLE_INDEXED_DATABASE
ENABLE_LINK_PREFETCH
ENABLE_GAMEPAD_DEPRECATED
```
Some of them cut dependencies. Note that arbitrary combinations of build options might result in compilation errors, and you should make sure that you know what option is doing before disabling it. When reporting issues, please check if it reproduces in the full version first (disabling WebKit2 and ENABLE_VIDEO is allowed though)

**Comment by annulen** [comment]
You might also need updated libraries from https://github.com/annulen/webkit/issues/789

**Comment by kaamui** [comment]
I couldn't go on building my app with qt 5.12.3 because of a Qt bug related to QWebView in ui forms that is not present in qt 5.9.5 

https://forum.qt.io/topic/103283/no-rule-to-make-target-qtwebkit-qwebview-qt-5-12-3-macx
https://stackoverflow.com/questions/41049586/referencing-webkit-or-webengine-in-ui-file-depending-on-qt-version

**Comment by annulen** [comment]
Workaround fir your ui file issue is to insert qwebview into form layout from c++ code

**Comment by kaamui** [comment]
I know but I am searching for the most effortless compromise as I have 3 platforms to manage and I really had no difficulties with the qt 5.9.5 on MacOS, and it is the version available in official repository of Ubuntu 18 which also simplify building of OpenBoard. With this choice, I only have to find a solution for the NPAPI support on MacOS that seems to be missing in qtwebkit's current state (we're thinking of removing swf support and replace it with HTML5 if making NPAPI work on MacOS isn't possible or isn't a good idea ... another point where I don't know what to think at this time).

**Comment by annulen** [comment]
It shouldn't be hard to restore npapi support on macos, someone just needs ti do the work. Nobody complained about it so far

**Comment by kaamui** [comment]
Ok, I'll try to look at it... but still, not sure if NPAPI support should be maintained at all, as it seems to not be the global trend, since 2013 : 

https://en.wikipedia.org/wiki/NPAPI#Support/deprecation

**Comment by annulen** [comment]
Same here. If you are using QtWebKit for trusted content, and probably with custom npapi plugins, security side of plugins should be of less concern for you, and your decision should depend on your requirements of compatibility with plugins which are currently in use in your application

**Comment by annulen** [comment]
Plaese try building with https://github.com/annulen/webkit/pull/820 - it should make NPAPI working on Mac

**Comment by kaamui** [comment]
Hello,

I finally have time to try this I'll let you know if it works, but before starting anything, I noticed in the diff viewer that an include was added in Source/Webkit/win/Plugins/PluginDatabase.cpp where I guess it was intended to be edited in Source/Webkit/mac/... ?

**Comment by annulen** [comment]
>Source/Webkit/win/Plugins/PluginDatabase.cpp

This file is platform independent, but is used only for windows in upstream
